cloudinit/sources/DataSourceVMware.py

```diff
@@ -507,46 +543,70 @@ def guestinfo_envvar_get_value(key):
     return handle_returned_guestinfo_val(key, os.environ.get(env_key, ""))
 
 
-def guestinfo(key, vmware_rpctool=VMWARE_RPCTOOL):
+def exec_vmware_rpctool(rpctool, args):
+    cmd = [rpctool]
+    cmd.append(args)
+    (stdout, stderr) = subp(cmd)
+    return (stdout, stderr)
+
+
+def exec_vmtoolsd(rpctool, args):
+    cmd = [rpctool, "--cmd"]
+    cmd.append(args)
+    (stdout, stderr) = subp(cmd)
+    return (stdout, stderr)
+
+
+def exec_rpctool(rpctool, args):
+    if is_rpctool_vmtoolsd(rpctool):
+        return exec_vmtoolsd(rpctool, args)
+    return exec_vmware_rpctool(rpctool, args)
+
+
+def guestinfo(key, rpctool):
     """
     guestinfo returns the guestinfo value for the provided key, decoding
     the value when required
     """
-    val = guestinfo_get_value(key, vmware_rpctool)
+    val = guestinfo_get_value(key, rpctool)
     if not val:
         return None
-    enc_type = guestinfo_get_value(key + ".encoding", vmware_rpctool)
+    enc_type = guestinfo_get_value(key + ".encoding", rpctool)
     return decode(get_guestinfo_key_name(key), enc_type, val)
 
 
-def guestinfo_get_value(key, vmware_rpctool=VMWARE_RPCTOOL):
+def guestinfo_get_value(key, rpctool):
     """
     Returns a guestinfo value for the specified key.
     """
-    LOG.debug("Getting guestinfo value for key %s", key)
+    LOG.debug("Getting guestinfo value for key %s with %s", key, rpctool)
+
+    args = ["info-get " + get_guestinfo_key_name(key)]
 
     try:
-        (stdout, stderr) = subp(
-            [
-                vmware_rpctool,
-                "info-get " + get_guestinfo_key_name(key),
-            ]
-        )
+        (stdout, stderr) = exec_rpctool(rpctool, args)
         if stderr == NOVAL:
             LOG.debug("No value found for key %s", key)
         elif not stdout:
             LOG.error("Failed to get guestinfo value for key %s", key)
         return handle_returned_guestinfo_val(key, stdout)
     except ProcessExecutionError as error:
+        # No matter the tool used to access the data, if NOVAL was returned on
+        # stderr, do not raise an exception.
         if error.stderr == NOVAL:
             LOG.debug("No value found for key %s", key)
         else:
+            # Any other result gets logged as an error, and if the tool was
+            # vmware-rpctool, then raise the exception so the caller can try
+            # again with vmtoolsd.
             util.logexc(
                 LOG,
                 "Failed to get guestinfo value for key %s: %s",
                 key,
                 error,
             )
+            if not is_rpctool_vmtoolsd(rpctool):
+                raise error
```

Comment:
There's a few places where this function is called but there's no fallback logic. That would result in an exception where previously it would just return None.

I'm also not sure it makes sense to check every single call. Presumably, if rpctool is going to fail for one call, it would fail for every call, correct? Does a call exist to get the version or something like that? If call succeeds, use rpctool everywhere. If not, use vmtoolsd. If so, we could do that during class initialization to figure out which tool to use. Then for any future calls, we could stick with one tool.

Reply:
Sure, but I view the logic as low-risk for the way it is organized. It only decides on which tool to use once, but to do what you're suggesting would essentially still require the use of an if/then, pointing to which function to call each time. Either way there's an if/then.

Reply:
> It only decides on which tool to use once

It's possible I haven't grokked the datasource enough, but I think `guesteinfo_set_value` is a better example. It still has the raise exception in it, but has 4 different calls, 2 in `advertise_local_ip_addrs`, and 2 in `guestinfo_redact_keys`. I don't see exception handling around either of those, so what happens if we raise the exception? Rather than tracing through every possible exception path, I think it's easier to set it once instead of possibly falling back multiple times.

Reply:
It won't be raised multiple times. I could remove it from setinfo and probably should. Once `self.rpctool` is set to `vmtoolsd` after the first exception, the only way it's raised again is if it is unrelated to fallback.

Reply:
I see what you are saying. The other exception handlers shouldn't have to care about this decision. I guess I was getting a little cute with the way it was handled. Will refactor. Thanks!